### PR TITLE
Add basic action-based editor, minorly improved from vello_editor.

### PR DESCRIPTION
**Pull request by xorgy**

### CODE REVIEW

**Comment by xorgy** [comment]
This inherits some structural issues from vello_editor in its current state, and with `build_into(...)` it can't be `no_std`

**Review by waywardmonkeys — COMMENTED** [review]
Didn't review this really, just noticed a couple of nits.

**Review by nicoburns — CHANGES_REQUESTED** [review]
This generally looks great but I have a few suggestions.

File: parley/src/layout/editor.rs [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+    SetText(Arc<str>),
+    SetWidth(f32),
+    SetScale(f32),
+    SetDefaultStyle(Arc<[StyleProperty<'a, T>]>),
+    InsertOrReplaceCurrentSelection(Arc<str>),
+    DeleteCurrentSelection,
+    DeleteCurrentSelectionOrNextCluster,
+    DeleteCurrentSelectionOrToNextWordBoundary,
+    DeleteCurrentSelectionOrPreviousCluster,
+    DeleteCurrentSelectionOrToPreviousWordBoundary,
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:52]: I think I agree.

File: parley/src/layout/editor.rs [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+#[non_exhaustive]
+pub enum PlainEditorOp<'a, T>
+where
+    T: Brush + Clone + Debug + PartialEq + Default,
+{
+    SetText(Arc<str>),
+    SetWidth(f32),
+    SetScale(f32),
+    SetDefaultStyle(Arc<[StyleProperty<'a, T>]>),
+    InsertOrReplaceCurrentSelection(Arc<str>),
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:47]: It used to in an intermediate version of this, will change it back. I would like to eventually ditch the named lifetime altogether.

File: parley/src/layout/mod.rs [hunk 3]
```diff
@@ -12,6 +12,11 @@ pub(crate) mod data;
 
 pub mod cursor;
 
+// TODO: make editor `no_std` capable.
+//       `std` required only because of `RangedEditor::build_into`.
+#[cfg(feature = "std")]
+pub mod editor;
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/mod.rs:18]: It is very closely tied to the layout, so I think it belongs at least in layout.

File: parley/src/lib.rs [hunk 4]
```diff
@@ -97,6 +97,7 @@ mod util;
 pub mod layout;
 pub mod style;
 
+pub use peniko::kurbo::Rect;
 pub use peniko::Font;
 
 pub use builder::{RangedBuilder, TreeBuilder};
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/lib.rs:103]: That's a fair reason to have a bare re-export.

File: parley/src/layout/editor.rs [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    buffer: String,
+    layout: Layout<T>,
+    selection: Selection,
+    cursor_mode: VisualMode,
+    width: f32,
+    scale: f32,
+}
+
+#[non_exhaustive]
+pub enum PlainEditorOp<'a, T>
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:41]: I've added some basic doc comments for the public APIs.

File: parley/src/layout/editor.rs [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+use alloc::{sync::Arc, vec::Vec};
+
+#[derive(Copy, Clone, Debug)]
+pub enum ActiveText<'a> {
+    FocusedCluster(Affinity, &'a str),
+    Selection(&'a str),
+}
+
+#[derive(Default)]
+pub struct PlainEditor<'a, T>
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:26]: Added.

File: parley/src/layout/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/mod.rs:18]: I've reexported `PlainEditor` and `PlainEditorOp` on Parley, I'm convinced by your argument for that.

File: parley/src/layout/editor.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs]: I've made these operation names simpler.

File: parley/src/layout/editor.rs [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+where
+    T: Brush + Clone + Debug + PartialEq + Default,
+{
+    font_cx: FontContext,
+    layout_cx: LayoutContext<T>,
+    default_style: Arc<[StyleProperty<'a, T>]>,
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:29]: Having reviewed the options for this, keeping things as they are is probably the right way to go. The next iteration ideally should make both the editor and the operations free of named lifetimes.

[the remark this answers is not in the document]
Comment (nicoburns) [reply parley/src/layout/editor.rs:29]: @xorgy Have you made this work in Masonry? I feel like this API is likely to be unusable for setting styles that contain lifetimes (without either leaking or hardcoding them (so they have a `'static` lifetime)). My suggestion in the short term would be create duplicate "owned" variants of `StyleProperty` (and sub-structs).

**Review by nicoburns — APPROVED** [review]
I still have concerns around the styling API, but happy for those to be addressed in a follow up (or to be shown why they are not a problem in practice).

File: examples/vello_editor/src/main.rs [hunk 8]
```diff
@@ -189,13 +205,12 @@ fn main() -> Result<()> {
...
     };
 
-    app.editor.set_text(text::LOREM);
-
     // Create and run a winit event loop
     let event_loop = EventLoop::new()?;
     event_loop
         .run_app(&mut app)
         .expect("Couldn't run event loop");
+    print!("{}", app.editor.text());
```
Comment (raphlinus) [inline examples/vello_editor/src/main.rs:213]: Was it intended to leave this in?

File: parley/src/layout/editor.rs [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+impl<'a, T> PlainEditor<'a, T>
+where
+    T: Brush + Clone + Debug + PartialEq + Default,
+{
+    /// Run a series of `PlainEditorOp`s, updating the layout if necessary
+    pub fn transact(
+        &mut self,
+        font_cx: &mut FontContext,
+        layout_cx: &mut LayoutContext<T>,
+        t: impl IntoIterator<Item = PlainEditorOp<'a, T>>,
```
[the remark this answers is not in the document]
Comment (DJMcNab) [reply parley/src/layout/editor.rs:145]: I see. I don't think we can possibly avoid doing this, so I think this is the right API. 
Some operations could be applied immediately, but it's unclear if that's worthwhile.

[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:145]: I think that invalidation is the correct final form, but operations were an expedient way to do this. Taking configuration (style, width, and scale) out of the operation set is probably a good first move. PlainEditor should probably be constructed with an owned default style, a width, and a scale; and then each of these should be mutable with a method.

File: parley/src/layout/editor.rs [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+    ) {
+        let mut layout_after = false;
+
+        for op in t.into_iter() {
+            match op {
+                PlainEditorOp::SetText(is) => {
+                    self.buffer.clear();
+                    self.buffer.push_str(&is);
+                    layout_after = true;
```
[the remark this answers is not in the document]
Comment (xorgy) [reply parley/src/layout/editor.rs:154]: Becomes invalid, it's a bug.

**Pull request merged**